### ISSUE TITLE
SteamOS 3.6..0 Preview uses zram, breaks CryoUtilities swap management

### ISSUE THREAD

**Issue by omor1** [CryoByte33/steam-deck-utilities#179]
SteamOS 3.6.0 uses zram; since this has higher priority than the swap file, this breaks `getSwapFileLocation`.

It's a bit unclear to me if zram is being used instead of, or in addition to the swap file.
This might also impact recommended swappiness: since swap is in compressed memory, it's much cheaper to write to/read from and it might mean it's better to proactively push less frequently used memory to zram.

**Comment by CryoByte33** [comment]
Good catch, I haven't had the time to try 3.6 preview, but it was working on `main` just a month or 2 ago, so Valve must have made another change. I have an update nearly ready for CU so I'll try to batch this in.

As for swappiness, it shouldn't make that much of a difference since raw RAM and ZRAM will both operate at similar speeds, but ZRAM will benefit from additional compression. As long as the swap devices have the correct precedence, you should see similar performance, but as always I will have to test to make sure.

**Comment by CryoByte33** [comment]
As an aside, I recently got a new job and have been very busy, sorry for what seems like no work on this project, I've just been doing a lot on a private testing repo since I made a lot of breaking changes :)

**Comment by krozgrov** [comment]
Does that mean to uninstall if upgrading to 3.6.0?  Or can we keep cryo installed just knowing that zram is being used?

**Comment by CryoByte33** [comment]
> Does that mean to uninstall if upgrading to 3.6.0? Or can we keep cryo installed just knowing that zram is being used?

CU should still be safe on 3.6, just avoid the swap resize function until a permanent fix is out. Aside from that you should be set, to the best of my knowledge.

**Comment by Aepoh** [comment]
Came here to make this issue, glad to see someone else got to it first.
![image](https://github.com/CryoByte33/steam-deck-utilities/assets/89321767/6788c0a4-8832-4213-8083-074677fde948)
I don't really know what priority means, but I assume it relates to Swappiness. If that's the case and zram is being prioritized then it would probably break the Swappiness setting as well.

**Comment by omor1** [comment]
> I don't really know what priority means, but I assume it relates to Swappiness. If that's the case and zram is being prioritized then it would probably break the Swappiness setting as well.

Priority is used by the Linux kernel to determine where to swap out pages evicted from memory—it generally tries using higher-priority swap locations before lower-priority ones. In this case, it tries evicting pages into zram (compressed memory) before it uses the swap file.

It is entirely orthogonal to swappiness, which determines how proactive the kernel is about evicting pages from memory.

**Comment by Aepoh** [comment]
> > I don't really know what priority means, but I assume it relates to Swappiness. If that's the case and zram is being prioritized then it would probably break the Swappiness setting as well.
> 
> 
> 
> Priority is used by the Linux kernel to determine where to swap out pages evicted from memory—it generally tries using higher-priority swap locations before lower-priority ones. In this case, it tries evicting pages into zram (compressed memory) before it uses the swap file.
> 
> 
> 
> It is entirely orthogonal to swappiness, which determines how proactive the kernel is about evicting pages from memory.

Ok so not the same thing, gotcha. I wonder if having two page files like that is problematic. Like if we imagine the page files as buckets and the computer is dumping data into the first bucket and then needs to dump data into the second one because the first one is full, is that going to be a seamless process or will there be some hitching/slowdown/stutter involved?

**Comment by grantdocet** [comment]
Good morning, now that VALVE has released OS 3.6 when you think to update the cryoutilities? thanks
Luke

**Comment by cheako** [comment]
I don't get using zram...  It creates a compressed virtual block device and then you set it up like a swap partition.  I feel zswap is much better, design wise.

Firstly because of the implications of overcommit_memory(0) and overcommit_ratio(50), I don't feel like going to go into it but the TLDR is that overcommit_ratio is not that at all in mode 0...  It's the kernel reserve amount.  In other words, by default the linux kernel will occupy %50 of ram+swap for its self.  Thus making swap allocation essential, as otherwise applications will not have access to the full ram.  The old advice of 2x ram as swap, comes from these defaults and is still an essential practice(because 1/2 > 1/3).  ram -** swap <--- 1/3 for ram, with 2x as swap.

Secondly because it solves the issue of SSD wear, by buffering swap writes into an internal compressed ram buffer.  Plus has the added benefit that when SSD writes occur, they are smaller as having been compressed.

So, I think tools optimizing the steamdeck...  Should create and enable a swapfile, enable zswap, then remove the default zram brokenness.
